esp-wifi/src/wifi/event.rs

```diff
@@ -0,0 +1,325 @@
+use alloc::boxed::Box;
+use core::cell::RefCell;
+
+use critical_section::Mutex;
+
+use super::WifiEvent;
+
+pub(crate) mod sealed {
+    use super::*;
+
+    pub trait Event {
+        /// Get the static reference to the handler for this event.
+        fn get_handler() -> &'static Mutex<RefCell<Option<Box<Handler<Self>>>>>;
+        /// # Safety
+        /// `ptr` must be a valid for casting to this event's inner event data.
+        unsafe fn from_raw_event_data(ptr: *mut crate::binary::c_types::c_void) -> Self;
+    }
+}
+/// The type of handlers of events.
+pub type Handler<T> = dyn FnMut(&T) + Sync + Send;
+
+fn default_handler<Event: 'static>() -> Box<Handler<Event>> {
+    fn drop_ref<T>(_: &T) {}
+    Box::new(drop_ref)
+}
+
+/// Extension trait for setting handlers for an event.
+///
+/// Register a new event handler like:
+/// ```
+/// # use esp_wifi::wifi::event::{self, *};
+/// # fn new_handler(_: &ApStaconnected) {}
+/// event::ApStaconnected::update_handler(|prev, event| {
+///     prev(event);
+///     new_handler(event);
+/// })
+/// ```
+// Implemented like this instead of free functions because the syntax would be
+// ```
+// event::update_handler::<event::ApStaconnected, _>(...)
+// ```
+pub trait EventExt: sealed::Event + Sized + 'static {
+    /// Get the handler for this event, replacing it with the default handler.
+    fn take_handler() -> Box<Handler<Self>> {
+        critical_section::with(|cs| {
+            Self::get_handler()
+                .borrow_ref_mut(cs)
+                .take()
+                .unwrap_or(default_handler::<Self>())
+        })
+    }
+    /// Set the handler for this event, returning the old handler.
+    fn replace_handler<F: FnMut(&Self) + Sync + Send + 'static>(f: F) -> Box<Handler<Self>> {
+        critical_section::with(|cs| {
+            Self::get_handler()
+                .borrow_ref_mut(cs)
+                .replace(Box::new(f))
+                .unwrap_or(default_handler::<Self>())
+        })
+    }
+    /// Atomic combination of [`take_handler`] and [`replace_handler`]. Use this
+    /// to add a new handler while preserving the previous.
+    fn update_handler<F: FnMut(&mut Handler<Self>, &Self) + Sync + Send + 'static>(mut f: F) {
+        critical_section::with(move |cs| {
+            let mut handler: Box<Handler<Self>> = Self::take_handler();
+            Self::get_handler()
+                .borrow_ref_mut(cs)
+                .replace(Box::new(move |event| f(&mut handler, event)));
```

Comment:
If the intent is for the user to always call the previous handler (the function's docs comments implies this), you can call the old handler in this closure, and not pass it to the user.